### PR TITLE
Reading arclink XML to ObsPy Inventory

### PR DESCRIPTION
**Pull request by Jollyfant**
Initial work on importing inventoryXML to ObsPy Inventory. Opened a pull request so we can track the changes (#1536).

Modified SC3ML script to read arclink XML. The formats are very similar and usually just swap around XML `<tags>` with `="attributes"`.

```
from obspy import read_inventory
inv = inventory_read("Package_1475048499259-GFZ_73152721_inventory.sc3.xml", format="INVENTORYXML")
```
- [x]  Refactor and tidy up; make more robust
- [x]  PEP8
- [x]  Add tests

### CODE REVIEW

**Comment by megies** [comment]
Cool thing that you're attacking this one! :+1: 

I'll add this to the 1.1.0 milestone instead. If it's not ready when we decide on releasing 1.1.0 we can still bump it.

**Comment by megies** [comment]
The one stray test fail is unrelated. (http://tests.obspy.org/52432/).

[One element type is not covered by tests](https://github.com/obspy/obspy/pull/1539/files#diff-0c3abd011149c1aec8d692a846c1e6b2R783), think you can dig up one of these and include for testing, by any chance?
```python
 +    elif(elem_type == 'responsePolynomial'): 
```

In case this is merged before [this](https://github.com/obspy/obspy/pull/1539/files#diff-0c3abd011149c1aec8d692a846c1e6b2R91) is fixed, please open up a follow-up issue. Also, I'd prefer to have the first line of the routine to be..
```python
raise NotImplementedError("Not implemented due to missing schema file.")
```
... or something similar, so that it's clear what the problem is when an error pops up (not everybody might right away see the note in the docstring..).

**Comment by Jollyfant** [comment]
I haven't been able to find data with `responsePolynomial`. But I'll do some more digging. This pull is in no rush to be merged IMHO.

**Comment by Jollyfant** [comment]
The schema was provided by Andres. Won't do the modification of `client.arclink` in this pull though, so removing it from the list..

File: obspy/io/arclink/inventory.py [hunk 1]
```diff
@@ -0,0 +1,908 @@
...
+
+def _is_inventory_xml(path_or_file_object):
+    """
+    Simple function checking if the passed object contains a valid arclink XML
+    1.0 file. Returns True of False.
+    The test is not exhaustive - it only checks the root tag but that should
+    be good enough for most real world use cases. If the schema is used to
+    test for a StationXML file, many real world files are false negatives as
+    they don't adhere to the standard.
+    :param path_or_file_object: File name or file like object.
```
Comment (megies) [inline obspy/io/arclink/inventory.py]: Please put a blank line before `params` block. In general we also have a blank line after the first line (in this case after the first two lines) of the docstring.

File: obspy/io/arclink/inventory.py [hunk 2]
```diff
@@ -0,0 +1,908 @@
...
+        "decimation_correction": decimation_corr
+    }
+
+    # Different processing for different types of responses
+    # currently supported:
+    # PAZ
+    # COEFF
+    # FIR
+    # Polynomial response is not supported, could not find example
+    if(elem_type == 'responsePAZ'):
```
Comment (megies) [inline obspy/io/arclink/inventory.py]: Space after the `if` and remove obsolete brackets, please.

File: obspy/io/arclink/inventory.py [hunk 3]
```diff
@@ -0,0 +1,908 @@
...
+
+    elif(elem_type == 'datalogger'):
+        cf_transfer_function_type = "DIGITAL"
+        numerator = []
+        denominator = []
+        return CoefficientsTypeResponseStage(
+            cf_transfer_function_type=cf_transfer_function_type,
+            numerator=numerator, denominator=denominator, **kwargs)
+
+    elif(elem_type == 'responsePolynomial'):
```
Comment (megies) [inline obspy/io/arclink/inventory.py]: Space after the `if` and remove obsolete brackets, please.

File: obspy/io/arclink/inventory.py [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+        "decimation_delay": decimation_delay,
+        "decimation_correction": decimation_corr
+    }
+
+    # Different processing for different types of responses
+    # currently supported:
+    # PAZ
+    # COEFF
+    # FIR
+    # Polynomial response is not supported, could not find example
```
Comment (megies) [inline obspy/io/arclink/inventory.py]: Seems like it's implemented below alright..

Also, I think the `if/elif` clause below could use an..

```python
else:
    msg = "Unsupported response type (%s). Please contact developers." % elem_type
    warnings.warn(msg)
```

File: obspy/io/arclink/inventory.py [hunk 5]
```diff
@@ -0,0 +1,908 @@
...
+
+        # Write the FIR symmetry to what ObsPy expects
+        # A = NONE, B = ODD, C = EVEN
+        symmetry = _attr2obj(stage, "symmetry", str)
+        if(symmetry == 'A'):
+            symmetry = 'NONE'
+        elif(symmetry == 'B'):
+            symmetry = 'ODD'
+        elif(symmetry == 'C'):
+            symmetry = 'EVEN'
```
Comment (megies) [inline obspy/io/arclink/inventory.py]: Again, please get rid of those brackets..

File: obspy/io/arclink/tests/test_inventory_xml.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import inspect
+import os
+import unittest
+
+from obspy.core.inventory import read_inventory
+from obspy.io.arclink.inventory import validate_arclink_xml
+
+
+class SC3MLTestCase(unittest.TestCase):
```
Comment (megies) [inline obspy/io/arclink/tests/test_inventory_xml.py]: Just a nit-pick.. but this should probably read `InventoryXMLTestCase` (also change at the bottom, if you change it)

File: obspy/io/arclink/tests/test_inventory_xml.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        Read example stationXML/sc3ml format to Inventory
+        """
+        self.data_dir = os.path.join(os.path.dirname(os.path.abspath(
+            inspect.getfile(inspect.currentframe()))), "data")
+        self.arclink_xml_path = os.path.join(self.data_dir,
+                                             "arclink_inventory.xml")
+        self.station_xml_path = os.path.join(self.data_dir, "gols_station.xml")
+
+    def test_validate_inventory_against_schema(self):
+        self.assertTrue(validate_arclink_xml(self.arclink_xml_path))
```
Comment (megies) [inline obspy/io/arclink/tests/test_inventory_xml.py]: Maybe add a validation of the station XML file that should return `False`?

**Comment by megies** [comment]
Please see https://github.com/obspy/obspy/wiki/How-to%3A-add-a-new-submodule. You need to make some changes for the new module to appear in docs properly.

I'll trigger a docs build for this PR, so you'll be able to check the docs without building yourself.. +DOCS

**Comment by Jollyfant** [comment]
I have tried converting a fdsnXML from IRIS that includes a polynomial to an SC3ML and Arclink inventory for testing:

    http://service.iris.edu/fdsnws/station/1/query?network=IU&level=response&station=ANTO&location=30&channel=LDO

Running it through `seiscomp exec fdsnxml2inv` throws a warning and doesn't save the response to the outfile:

    Processing ./irisinput.xml
     - parsing StationXML
    11:52:33 [warning] InstrumentSensitivity: missing mandatory attributes: Frequency, Value
    11:52:33 [warning] Stage: missing mandatory attribute: StageGain
     - converting into SeisComP-XML
    Finished processing

For good measure I tried `staxml2inv` too but no cigar.

@andresh, @krischer -- help?

**Comment by megies** [comment]
> @andres-h, @krischer -- help?

also CCing @jsaul, he might also be able to give feedback

**Comment by andres-h** [comment]
On 10/06/2016 02:02 PM, Jollyfant wrote:
> I have tried converting a fdsnXML from IRIS that includes a polynomial
> to an SC3ML and Arclink inventory for testing:
>
> |http://service.iris.edu/fdsnws/station/1/query?network=IU&level=response&station=ANTO&location=30&channel=LDO
> |
>
> Running it through |seiscomp exec fdsnxml2inv| throws a warning and
> doesn't save the response to the outfile:
>
> |Processing ./irisinput.xml - parsing StationXML 11:52:33 [warning]
> InstrumentSensitivity: missing mandatory attributes: Frequency, Value
> 11:52:33 [warning] Stage: missing mandatory attribute: StageGain -
> converting into SeisComP-XML Finished processing |

The input is not correct StationXML as it doesn't validate against 
https://raw.githubusercontent.com/FDSN/StationXML/v1.0/fdsn-station.xsd

$ xmllint --schema fdsn-station.xsd 
'http://service.iris.edu/fdsnws/station/1/query?network=IU&level=response&station=ANTO&location=30&channel=LDO' 
--noout
http://service.iris.edu/fdsnws/station/1/query?network=IU&level=response&station=ANTO&location=30&channel=LDO:74: 
element InputUnits: Schemas validity error : Element 
'{http://www.fdsn.org/xml/station/1}InputUnits': This element is not 
expected. Expected is ( {http://www.fdsn.org/xml/station/1}Value ).
http://service.iris.edu/fdsnws/station/1/query?network=IU&level=response&station=ANTO&location=30&channel=LDO:83: 
element Stage: Schemas validity error : Element 
'{http://www.fdsn.org/xml/station/1}Stage': Missing child element(s). 
Expected is one of ( {http://www.fdsn.org/xml/station/1}Decimation, 
{http://www.fdsn.org/xml/station/1}StageGain ).
http://service.iris.edu/fdsnws/station/1/query?network=IU&level=response&station=ANTO&location=30&channel=LDO 
fails to validate


> For good measure I tried |staxml2inv| too but no cigar.

staxml2inv implements an old draft of StationXML and should be no longer 
used.


Regards,
Andres.

**Comment by Jollyfant** [comment]
Thanks Andres, that is a fair assement. But why is IRIS returning invalid stationXML >_>

Anyway, I don't think `PolynomialResponseStage` is tested in io.stationXML either.

**Comment by krischer** [comment]
`io.stationxml` does test (de)serialization of polynomial stages (I'm not sure we also test the actual response calculation but we probably should). For this we just use an autogenerated "complete" stationxml file: https://github.com/obspy/obspy/blob/master/obspy/io/stationxml/tests/data/full_random_stationxml.xml

We generated it from the schema - the same should be possible with the arclink schema. I don't remember which tool we used but there should be a number of different ones available.

**Comment by megies** [comment]
@Jollyfant what's the status on this PR? Reading through the conversation it's probably ready? If it becomes ready within 2-3 weeks it'll make it into 1.1.0.

**Comment by Jollyfant** [comment]
The code works but the polynomial response stage remains untested but the code works on the case that was submitted through the mailing list. I'll take a look tomorrow and see if we can merge this.

**Comment by megies** [comment]
If it's just the polynomial response stage, you can just drop in a `NotImplementedError` for the time being. It shouldn't hold up the rest..

**Comment by megies** [comment]
@Jollyfant, any news on this one? Like I said, if it's just the `PolynomialResponse` that's missing just drop in an `NotImplementedError` for that one. This will be bumped this to 1.2.0 in a week or so, if there's no updates.

**Comment by Jollyfant** [comment]
Ok! Implementing this now.

**Comment by Jollyfant** [comment]
Alright, rebased on master and force pushed. Waiting for integration tests to pass..

**Comment by Jollyfant** [comment]
@megies unrelated fails?

**Comment by megies** [comment]
> unrelated fails?

Yes, we're seeing those sporadic fails across all commits. There seems to be problems with requests to examples.obspy.org but we couldn't find the problem/reason yet..

**Comment by megies** [comment]
@Jollyfant, please rebase file '.k.swp' out of this PR.

**Comment by Jollyfant** [comment]
Woops, thanks. I missed that.

**Comment by megies** [comment]
It's OK, I was just adding the elemt type to the error message, but you did it too now, see https://github.com/obspy/obspy/pull/1539/files#diff-0c3abd011149c1aec8d692a846c1e6b2R706

Probably a NotImplementedError is better fitting there than a ValueError..

**Comment by megies** [comment]
> Probably a NotImplementedError is better fitting there though, than a ValueError..

@Jollyfant, can you change this one thing and rebase on current master to check CI?

**Comment by Jollyfant** [comment]
I thought I had replaced that ValueError to NotImplementedError before. Anyway rebased to master,  no conflicts!

**Comment by megies** [comment]
Great!

One last thing.. would you mind squashing the last commit into 169f336 so that we get rid of that `.swp` file completely?

**Comment by Jollyfant** [comment]
How would I go about doing that? Can I squash them during the pull merge?

**Comment by megies** [comment]
> How would I go about doing that? Can I squash them during the pull merge?

We could do that, but if there is granularity in your commit history of this PR that might be worth keeping.. that would be lost as well.

If you're working with git on the command line..  `$ git rebase -i master` .. and then an editor window will pop up, you can move that last commit up to second place and mark it as `fixup` (meld) (https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history, https://asciinema.org/a/9593)

If you're not comfortable with that, just leave it as is, we can do it at the end.

**Comment by megies** [comment]
ping @Jollyfant  we want to get 1.1.0 ready within days...

**Comment by Jollyfant** [comment]
This pull is pretty complete. I will rebase tomorrow and make sure it's ready.

**Comment by Jollyfant** [comment]
+DOCS

**Comment by Jollyfant** [comment]
@megies docsbot tried and died. I'll try to reset it with another

+DOCS

File: obspy/io/arclink/__init__.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+"""
+obspy.io.arclink - Arclink Inventory Files
+=============================================
```
Comment (krischer) [inline obspy/io/arclink/__init__.py]: A bit too long.

File: obspy/io/arclink/inventory.py [hunk 9]
```diff
@@ -0,0 +1,929 @@
...
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+
+import inspect
+import re
+import math
+import os
+import warnings
```
Comment (krischer) [inline obspy/io/arclink/inventory.py]: Please sort the stdlib imports alphabetically and move the obspy import down to the other obspy imports.

File: obspy/io/arclink/inventory.py [hunk 10]
```diff
@@ -0,0 +1,929 @@
...
+    function to read floattype to cls object (based on _read_floattype)
+    normally ObsPy would read this directly from a tag, but with different
+    tag names this is no longer possible; instead we just pass the value
+    and not the tag name. We always set the unit/datum/uncertainties to None
+    because they are not provided by sc3ml ?
+
+    :param elem: float value to be converted
+    :param cls: obspy.core.inventory class
+    """
+
```
Comment (krischer) [inline obspy/io/arclink/inventory.py]: unnecessary space

**Comment by megies** [comment]
> @megies docsbot tried and died. I'll try to reset it with another
> 
> +DOCS

It doesn't make a difference to put this magic multiple times, the docs buildbot only checks if the last commit has a docs buildbot commit status yet or not..

**Comment by megies** [comment]
As mentioned before, can you please squash this commit https://github.com/obspy/obspy/pull/1539/commits/cd8d114ef6a528a1fec10658ccb8b92f065abaae into this commit https://github.com/obspy/obspy/commit/bf9649ae7f759c0d2a8e9ecd49e6c97e99a0b3f0 so that this swp file gets purged from history completely?

**Comment by megies** [comment]
This commit https://github.com/Jollyfant/obspy/commit/7683558743195fdd467ce167b19c10dadcf998e9 should also be squashed into https://github.com/Jollyfant/obspy/commit/bf9649ae7f759c0d2a8e9ecd49e6c97e99a0b3f0.

**Comment by Jollyfant** [comment]
I'm not sure how to squash two particular commits. We can squash all of them in the pull merge.

**Comment by Jollyfant** [comment]
Merged with master to resolve conflicts with `datamark` => `win`.

@krischer - Yeah, I guess SC3ML is loosely based on arclinkXML. But one way or another we need to tell ObsPy where to find each fields in the tree.. not really a way to put them together IMHO.

**Comment by megies** [comment]
> I haven't been able to find data with responsePolynomial. But I'll do some more digging. This pull is in no rush to be merged IMHO.

Maybe we could drop in a warning message if the `responsePolynomial` switch is hit  (that it's untested) and tell people to supply a test file? Just an idea though..

**Comment by megies** [comment]
Might be good to cover this case in tests: https://codecov.io/gh/obspy/obspy/compare/878232ef52d361821be269f9653717ec5c6a5767...aedaa12872dd78ac1b173deb200d88657d037b70/diff#D12-597 as it looks non-trivial..

**Comment by megies** [comment]
Tell me when you're done than I'll do the various commit cleanups I mentioned before couple of times..

**Comment by Jollyfant** [comment]
@megies - it raises when responsePolynomial is hit to contact us so we can implement it.

    raise NotImplementedError("responsePolynomial not implemented. Contact the ObsPy developers")

I'm adding a test for analogue filter chains right now.

**Comment by Jollyfant** [comment]
CI fails appear unrelated. If you have time you can do the clean up.

**Comment by krischer** [comment]
Looks good to me. Can this be merged?

**Comment by megies** [comment]
Rebased, cleaned up history and forc-pushed (previous commit hash was b316619a648dd2576b8c74ece505671987e9a39c)

**Comment by Jollyfant** [comment]
All are happy except for Code Canary. There's some code behind the `raise` for `responsePolynomial.` that's causing it. Good to merge IMHO.

**Comment by krischer** [comment]
What do you mean with code canary?

**Comment by Jollyfant** [comment]
I mean codecov >_> it's avatar is a bird right?

**Comment by krischer** [comment]
Haha I see :-) I always thought its an umbrella - but who knows.

**Comment by Jollyfant** [comment]
Actually I think you are right and that's why it is called "coverage" huh. Anyway, thanks @megies and @krischer for the help.

**Pull request merged**